cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/helpers/functional/BasicLattices.kt

```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright (c) 2024, Fraunhofer AISEC. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *                    $$$$$$\  $$$$$$$\   $$$$$$\
+ *                   $$  __$$\ $$  __$$\ $$  __$$\
+ *                   $$ /  \__|$$ |  $$ |$$ /  \__|
+ *                   $$ |      $$$$$$$  |$$ |$$$$\
+ *                   $$ |      $$  ____/ $$ |\_$$ |
+ *                   $$ |  $$\ $$ |      $$ |  $$ |
+ *                   \$$$$$   |$$ |      \$$$$$   |
+ *                    \______/ \__|       \______/
+ *
+ */
+package de.fraunhofer.aisec.cpg.helpers.functional
+
+import de.fraunhofer.aisec.cpg.helpers.IdentitySet
+import de.fraunhofer.aisec.cpg.helpers.identitySetOf
+import de.fraunhofer.aisec.cpg.helpers.toIdentitySet
+import java.util.IdentityHashMap
+import kotlin.Pair
+import kotlin.collections.component1
+import kotlin.collections.component2
+import kotlin.collections.toMap
+
+/**
+ * A complete lattice is an ordered structure of values of type [T]. [T] could be anything, e.g., a
+ * set, a new data structure (like a range), or anything else. [T] depends on the analysis and
+ * typically has to abstract the value for the specific purpose.
+ *
+ * This class is actually used to hold individual instances of the lattice's elements and to compute
+ * bigger elements depending on these two elements.
+ *
+ * Implementations of this class have to implement the comparator, the least upper bound of two
+ * lattices.
+ */
+interface LatticeElement<T> : Comparable<LatticeElement<T>> {
+    val elements: T
```

Comment:
The wording is still confusing to me. Is `LatticeElement` now representing one element of the set (represented as the lattice) and if yes, why is its property also called `elements`?

From what I gather this really is the complete lattice, so maybe drop the `Element` from the class name? Can we constrain `T` to a collection?

Reply:
It's a single element inside the lattice. I refactored `elements` to `value` which may help a bit. I could probably also implement just individual functions for `lub` and the comparisons without having this wrapper class. However, I wouldn't store all possible elements a lattice can have because we would e.g. have to compute and (temporarily) store the powerset of all nodes in a function (for a lattice over a powerset) and, when it comes to the maps, even map all nodes to each element in the powerset. This simply requires too much memory and it's better to compute them on-demand.